array.c

```diff
@@ -7000,82 +7000,41 @@ rb_ary_permutation_size(VALUE ary, VALUE args, VALUE eobj)
 
 /*
  *  call-seq:
- *    array.permutation {|element| ... } -> self
- *    array.permutation(n) {|element| ... } -> self
- *    array.permutation -> new_enumerator
- *    array.permutation(n) -> new_enumerator
+ *    permutation(n = self.size) {|permutation| ... } -> self
+ *    permutation(n = self.size) -> new_enumerator
  *
- *  When invoked with a block, yield all permutations of elements of +self+; returns +self+.
- *  The order of permutations is indeterminate.
+ *  Iterates over permutations of the elements of +self+;
+ *  the order of permutations is indeterminate.
  *
- *  When a block and an in-range positive Integer argument +n+ (<tt>0 < n <= self.size</tt>)
- *  are given, calls the block with all +n+-tuple permutations of +self+.
+ *  The examples here use a helper method to gather and return permutations:
  *
- *  Example:
- *
- *    a = [0, 1, 2]
- *    a.permutation(2) {|permutation| p permutation }
- *
- *  Output:
- *
- *    [0, 1]
- *    [0, 2]
- *    [1, 0]
- *    [1, 2]
- *    [2, 0]
- *    [2, 1]
- *
- *  Another example:
- *
- *    a = [0, 1, 2]
- *    a.permutation(3) {|permutation| p permutation }
- *
- *  Output:
+ *    def get_perms(a, n)
+ *      perms = []
+ *      a.permutation(n) {|perm| perms.push(perm) }
+ *      perms
+ *    end
  *
- *    [0, 1, 2]
- *    [0, 2, 1]
- *    [1, 0, 2]
- *    [1, 2, 0]
- *    [2, 0, 1]
- *    [2, 1, 0]
+ *  With a block and an in-range positive integer argument +n+ (<tt>0 < n <= self.size</tt>) given,
+ *  calls the block with each +n+-tuple permutations of +self+;
+ *  returns +self+:
  *
- *  When +n+ is zero, calls the block once with a new empty +Array+:
+ *    get_perms(a, 1) # => [[0], [1], [2]]
+ *    get_perms(a, 2) # => [[0, 1], [0, 2], [1, 0], [1, 2], [2, 0], [2, 1]]
+ *    get_perms(a, 3) # => [[0, 1, 2], [0, 2, 1], [1, 0, 2], [1, 2, 0], [2, 0, 1], [2, 1, 0]]
```

Comment:
Alternatively, you can use `Enumerator#to_a` without needing to define a `get_perms` method:

```ruby
[1, 2, 3].permutation.to_a # => [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
```

Reply:
Yes, but that call does not have a block (as the examples require).

Is this better?

```
perms = []
[1, 2, 3].permutation(3) {|perm| perms.push(perm) }
perms # => [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
```

Reply:
It's true that it does not call the block, but IMO, it's a lot more readable than calling `get_perms`.

Reply:
IMO, the examples must call the block (which is the primary form for the call-seq).  I've removed `get_perms`, which I agree is sub-optimal.